src/HomogeneousCalorimeter_geo.cpp

```diff
@@ -467,60 +523,85 @@ static std::tuple<int, int> add_12surface_disk(Detector& desc, Assembly& env, xm
-    modPV.addPhysVolID("sector", sector_id).addPhysVolID("module", total_id);
-  }
+  int mid = 0, total_id = 0;
+  float half_modx = modSize.x() * 0.5, half_mody = modSize.y() * 0.5;
+  auto points = epic::geo::fillRectangles({half_modx, half_mody}, modSize.x(), modSize.y(), rmin, rmax, phimin, phimax);
+  for (auto& p : points)
+    {
+      Transform3D tr_local = RotationZYX(-15.*degree, 0.0, 0.0) * Translation3D(p.x(), p.y(), 0.0);
+      auto modPV = (has_envelope ? env_vol.placeVolume(modVol, tr_local) : env.placeVolume(modVol, tr_global * tr_local));
+      modPV.addPhysVolID("sector", sector_id).addPhysVolID("module", total_id);
+      total_id = id_begin + mid++;
+      // std::cout << total_id << " " << p.x() << " " << p.y() << std::endl;  // use for debug
+    }
+
+  // Add the modules manually in the gap [Outer]
+  //
+  const int add_N_mod_outer = 56;
+  float addX_outer[add_N_mod_outer] = {9.225, 11.275, 13.325, 15.375, 17.425, 19.475, 44.075, 46.125, 60.475, 60.475, 62.525, 62.525, 62.525, 62.525,  -9.225, -11.275, -13.325, -15.375, -17.425, -19.475, -44.075, -46.125, -60.475, -60.475, -62.525, -62.525, -62.525, -62.525,  -9.225, -11.275, -13.325, -15.375, -17.425, -19.475, -44.075, -46.125, -60.475, -60.475, -62.525, -62.525, -62.525, -62.525,  9.225, 11.275, 13.325, 15.375, 17.425, 19.475, 44.075, 46.125, 60.475, 60.475, 62.525, 62.525, 62.525, 62.525};
+  float addY_outer[add_N_mod_outer] = {62.525, 62.525, 62.525, 62.525, 60.475, 60.475, 46.125, 44.075, 19.475, 17.425, 15.375, 13.325, 11.275, 9.225,  62.525, 62.525, 62.525, 62.525, 60.475, 60.475, 46.125, 44.075, 19.475, 17.425, 15.375, 13.325, 11.275, 9.225,  -62.525, -62.525, -62.525, -62.525, -60.475, -60.475, -46.125, -44.075, -19.475, -17.425, -15.375, -13.325, -11.275, -9.225, -62.525, -62.525, -62.525, -62.525, -60.475, -60.475, -46.125, -44.075, -19.475, -17.425, -15.375, -13.325, -11.275, -9.225};
+
+  for (int im = 0; im < add_N_mod_outer; im++)
+    {
+      Transform3D add_local = RotationZYX(-15.*degree, 0.0, 0.0) * Translation3D(addX_outer[im] * cm, addY_outer[im] * cm, 0.0);
+      auto modPV = (has_envelope ? env_vol.placeVolume(modVol, add_local) : env.placeVolume(modVol, tr_global * add_local));
+      modPV.addPhysVolID("sector", sector_id).addPhysVolID("module", total_id);
+      total_id++;
+    }
+
+  // Add the modules manually in the gap [Inner]
+  //
+  const int add_N_mod_inner = 36;
+  double addX_inner[add_N_mod_inner] = {1.025,1.025,3.075,3.075,5.125,7.175,7.175,9.225,9.225, -1.025,-1.025,-3.075,-3.075,-5.125,-7.175,-7.175,-9.225,-9.225, -1.025,-1.025,-3.075,-3.075,-5.125,-7.175,-7.175,-9.225,-9.225, 1.025,1.025,3.075,3.075,5.125,7.175,7.175,9.225,9.225};
+  double addY_inner[add_N_mod_inner] = {7.175,9.225,7.175,9.225,7.175,7.175,5.125,1.025,3.075, 7.175,9.225,7.175,9.225,7.175,7.175,5.125,1.025,3.075, -7.175,-9.225,-7.175,-9.225,-7.175,-7.175,-5.125,-1.025,-3.075, -7.175,-9.225,-7.175,-9.225,-7.175,-7.175,-5.125,-1.025,-3.075};
```

Comment:
Remove hardcoded positions. Avoid arrays and use vectors instead.

Reply:
I remove the hardcoded positions and read them by vectors